File: Components/DevExpressVCL/Demos/ExpressQuantumGrid/CBuilder/UnboundListDemo/UnboundListDemoMain.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#include "shellapi.hpp"
#pragma hdrstop

#include "UnboundListDemoMain.h"
#include "cxDataStorage.hpp"
#include "AboutDemoForm.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "cxClasses"
#pragma link "cxControls"
#pragma link "cxCustomData"
#pragma link "cxData"
#pragma link "cxDBData"
#pragma link "cxEdit"
#pragma link "cxFilter"
#pragma link "cxGraphics"
#pragma link "cxGrid"
#pragma link "cxGridCustomTableView"
#pragma link "cxGridCustomView"
#pragma link "cxGridDBTableView"
#pragma link "cxGridLevel"
#pragma link "cxGridTableView"
#pragma link "cxStyles"
#pragma link "cxGridCustomPopupMenu"
#pragma link "cxGridPopupMenu"
#pragma link "cxDataStorage"
#pragma link "cxLookAndFeels"
#pragma link "cxGridCardView"
#pragma link "cxLookAndFeelPainters"
#pragma resource "*.dfm"
TUnboundListDemoMainForm *UnboundListDemoMainForm;
const String TabChar = "\t";

//---------------------------------------------------------------------------
__fastcall TUnboundListDemoMainForm::TUnboundListDemoMainForm(TComponent* Owner)
  : TfmBaseForm(Owner)
{
}
//---------------------------------------------------------------------------

void __fastcall TUnboundListDemoMainForm::FormCreate(TObject *Sender)
{
  CustomerList = new TCustomerList();
  CustomerDataSource = new TCustomerDataSource(CustomerList);
  CustomizeGrid();
}
//---------------------------------------------------------------------------

void __fastcall TUnboundListDemoMainForm::FormDestroy(TObject *Sender)
{
  delete CustomerList;
  delete CustomerDataSource;
}

//---------------------------------------------------------------------------
void __fastcall TUnboundListDemoMainForm::GenerateColumns()
{
  TcxGridColumn* GridColumn;
  tvCustomers->ClearItems();

  GridColumn = tvCustomers->CreateColumn();
  GridColumn->Caption = "ID";
  GridColumn->DataBinding->ValueTypeClass = (TcxValueTypeClass)__classid(TcxIntegerValueType);
  GridColumn->Width = 50;

  GridColumn = tvCustomers->CreateColumn();
  GridColumn->Caption = "Customer";
  GridColumn->DataBinding->ValueTypeClass = (TcxValueTypeClass)__classid(TcxStringValueType);
  GridColumn->Width = 200;

  GridColumn = tvCustomers->CreateColumn();
  GridColumn->Caption = "Company";
  GridColumn->DataBinding->ValueTypeClass = (TcxValueTypeClass)__classid(TcxStringValueType);
  GridColumn->Width = 200;

  tvCustomers->DataController->CustomDataSource = CustomerDataSource;
}

//---------------------------------------------------------------------------
void __fastcall TUnboundListDemoMainForm::LoadData()
{
  TCustomer* ACustomer;
  AnsiString s;
  TStringList* StringList = new TStringList();
  try{
    StringList->LoadFromFile("contacts.txt");
    for(int i = 0; i < StringList->Count; i++) {
      ACustomer = new TCustomer(CustomerList->NextID);
      s = StringList->Strings[i];
      ACustomer->Name = s.SubString(1, s.Pos(TabChar) - 1);
      ACustomer->Description = s.SubString(s.Pos(TabChar) + 1, s.Length());
      CustomerList->Add(ACustomer);
    }
  }
   __finally {
     delete StringList;
   }
  CustomerDataSource->DataChanged();
  tvCustomers->DataController->GotoFirst();
}

//---------------------------------------------------------------------------
void __fastcall TUnboundListDemoMainForm::CustomizeGrid()
{
  GenerateColumns();
  LoadData();
}

//---------------------------------------------------------------------------
void __fastcall TUnboundListDemoMainForm::SaveData()
{
  TCustomer* ACustomer;
  TStringList* StringList = new TStringList();
  try {
    for(int i = 0; i < CustomerList->Count; i++) {
       ACustomer = CustomerList->Customers[i];
       StringList->Add(ACustomer->Name + TabChar + ACustomer->Description);
    }
    StringList->SaveToFile("contacts.txt");
  }
  __finally {
    delete StringList;
  }
}

//---------------------------------------------------------------------------
void __fastcall TUnboundListDemoMainForm::FormCloseQuery(TObject *Sender,
      bool &CanClose)
{
  int i = -1;
  if (CustomerDataSource->Modified)
    i = MessageDlg("Do you want to save the changes ?", mtConfirmation, TMsgDlgButtons() <<mbYes << mbNo << mbCancel, 0);
  switch (i) {
    case mrYes: SaveData(); break;
    case mrCancel: CanClose = false; break;
  }
}
//---------------------------------------------------------------------------

File: Components/DevExpressVCL/Demos/ExpressQuantumGrid/Delphi/UnboundListDemo/UnboundListDemoMain.pas
unit UnboundListDemoMain;

interface

uses
  Windows, Messages, SysUtils, Classes, Controls,  Forms, Dialogs, cxStyles,
  cxCustomData, cxGraphics, cxFilter, cxData, cxEdit, cxGridTableView,
  Menus, cxGridLevel, cxGridCustomTableView, cxClasses,
  cxControls, cxGridCustomView, cxGrid, StdCtrls, ComCtrls,
  cxGridCustomPopupMenu, cxGridPopupMenu, cxDataStorage, cxLookAndFeels,
  BaseForm, cxLookAndFeelPainters, cxGridCardView;

type
  TUnboundListDemoMainForm = class(TfmBaseForm)
    lvCustomers: TcxGridLevel;
    cxGrid: TcxGrid;
    tvCustomers: TcxGridTableView;
    cxGridPopupMenu: TcxGridPopupMenu;
    procedure FormCreate(Sender: TObject);
    procedure FormDestroy(Sender: TObject);
    procedure FormCloseQuery(Sender: TObject; var CanClose: Boolean);
  private
    procedure CustomizeGrid;
    procedure GenerateColumns;
    procedure LoadData;
    procedure SaveData;
  protected
    procedure LookAndFeelChanged(Sender: TcxLookAndFeel;
      AChangedValues: TcxLookAndFeelValues); override;
  end;

var
  UnboundListDemoMainForm: TUnboundListDemoMainForm;

implementation

uses
  UnboundListDemoClasses, AboutDemoForm;

{$R *.dfm}

const
  TabChar = #9;

var
  CustomerList: TCustomerList;
  CustomerDataSource: TCustomerDataSource;

procedure TUnboundListDemoMainForm.CustomizeGrid;
begin
  GenerateColumns;
  LoadData;
end;

procedure TUnboundListDemoMainForm.GenerateColumns;
begin
  with tvCustomers as TcxGridTableView do
  begin
    ClearItems;

    with CreateColumn as TcxGridColumn do
    begin
      Caption := 'ID';
      DataBinding.ValueTypeClass := TcxIntegerValueType;
      Width := 50;
    end;

    with CreateColumn as TcxGridColumn do
    begin
      Caption := 'Customer';
      DataBinding.ValueTypeClass := TcxStringValueType;
      Width := 200;
    end;

    with CreateColumn as TcxGridColumn do
    begin
      Caption := 'Company';
      DataBinding.ValueTypeClass := TcxStringValueType;
      Width := 200;
    end;
  end;

  tvCustomers.DataController.CustomDataSource := CustomerDataSource;
end;

procedure TUnboundListDemoMainForm.LoadData;
var
  ACustomer: TCustomer;
  I: Integer;
  s: string;
begin
  with TStringList.Create do
  try
    LoadFromFile('contacts.txt');
    for I := 0 to Count - 1 do
    begin
      ACustomer := TCustomer.Create(CustomerList.NextID);
      s := Strings[I];
      ACustomer.Name := Copy(s, 1, Pos(TabChar, s) - 1);
      ACustomer.Description := Copy(s, Pos(TabChar, s) + 1, Length(s));
      CustomerList.Add(ACustomer);
    end;
  finally
    Free;
  end;
  CustomerDataSource.DataChanged;
  tvCustomers.DataController.GotoFirst;
end;

procedure TUnboundListDemoMainForm.SaveData;
var
  ACustomer: TCustomer;
  I: Integer;
begin
  with TStringList.Create do
  try
    for I := 0 to CustomerList.Count - 1 do
    begin
       ACustomer := CustomerList.Customers[I];
       Add(ACustomer.Name + TabChar + ACustomer.Description);
    end;
    SaveToFile('contacts.txt');
  finally
    Free;
  end;
end;

procedure TUnboundListDemoMainForm.FormCreate(Sender: TObject);
begin
  CustomerList := TCustomerList.Create;
  CustomerDataSource := TCustomerDataSource.Create(CustomerList);
  CustomizeGrid;
end;

procedure TUnboundListDemoMainForm.FormDestroy(Sender: TObject);
begin
  CustomerDataSource.Free;
  CustomerList.Free;
end;

procedure TUnboundListDemoMainForm.FormCloseQuery(Sender: TObject; var CanClose: Boolean);
var
  I: Integer;
begin
  I := -1;
  if CustomerDataSource.Modified then
    I := MessageDlg('Do you want to save the changes ?', mtConfirmation, [mbYes, mbNo, mbCancel], 0);
  case I of
    mrYes:
      SaveData;
    mrCancel:
      CanClose := False;
  end;
end;

procedure TUnboundListDemoMainForm.LookAndFeelChanged(
  Sender: TcxLookAndFeel; AChangedValues: TcxLookAndFeelValues);
begin
  inherited LookAndFeelChanged(Sender, AChangedValues);
  UpdateTableViewStyleSheet(tvCustomers);
end;

end.
